lib/afinter.c

```diff
@@ -496,3 +501,83 @@ afinter_global_deinit(void)
     }
   current_internal_source = NULL;
 }
+
+/****************************************************************************
+ * Live collection of internal log messages.
+ ****************************************************************************/
+
+AFInterLive
+afinter_start_live_collection(void)
+{
+  AFInterSource *self;
+
+  if (is_live_collection)
+    return AFINTER_LIVE_COLLECTION_RUNNING;
+
+  if (current_internal_source != NULL)
+    return AFINTER_INTERNAL_SRC_PRESENT;
+
+  self = g_new0(AFInterSource, 1);
+
+  g_static_mutex_lock(&internal_msg_lock);
+  current_internal_source = self;
+  g_static_mutex_unlock(&internal_msg_lock);
+
+  is_live_collection = TRUE;
+
+  return AFINTER_LIVE_COLLECTION_STARTED;
+}
+
+void
+afinter_reset_live_collection(void)
+{
+  afinter_stop_live_collection();
+
+  if (is_live_collection)
+    {
+      if (internal_msg_queue)
+        _release_internal_msg_queue();
+    }
+}
+
+void
+afinter_stop_live_collection(void)
+{
+  if (is_live_collection)
+    {
+      if (current_internal_source != NULL)
+        {
+          AFInterSource *self = current_internal_source;
+
+          g_static_mutex_lock(&internal_msg_lock);
+          current_internal_source = NULL;
+          g_static_mutex_unlock(&internal_msg_lock);
+
+          is_live_collection = FALSE;
+          g_free(self);
+        }
+    }
+}
+
+void
+afinter_get_collected_messages(GString *result)
+{
+  LogMessage *msg;
+  const gchar *msg_text;
+
+  g_static_mutex_lock(&internal_msg_lock);
+
+  while (!g_queue_is_empty(internal_msg_queue))
+    {
+      msg = g_queue_pop_head(internal_msg_queue);
```

Comment:
This message needs to be unreffed  using `log_msg_unref`.

Reply:
@furiel : okay, done.